### PR TITLE
Always Shutdown PG Before pgBackRest Repo Host

### PR DESCRIPTION
**Pull request by andrewlecuyer**
Ensures PG instances are always shutdown prior to shutting down the pgBackRest repo host, such as when `Shutdown` is set to `true` in the PostgresCluster spec or when shutting down the cluster gracefully prior to a major PG upgrade.  This ensures the pgBackRest dedicated repository host is available for the full lifetime of any/all running PG instances, and WAL can therefore always be pushed as needed.

Additionally, when setting `Shutdown` to `true` in the PostgresCluster spec in order to gracefully shutdown the cluster prior to a major PG upgrade, the `PostgresVersion` is now also set to the `FromPostgresVersion`.  This ensures the cluster continues to reconcile with the proper PG version (i.e. while the cluster is shutting down) prior to the upgrade completing.

**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?
   - [x] Have you added automated tests?

**Type of Changes:**

 
 - [ ] New feature
 - [x] Bug fix
 - [ ] Documentation
 - [ ] Testing enhancement
 - [ ] Other

**What is the current behavior (link to any open issues here)?**

- The pgBackRest dedicated repo host does not wait for all PG instances to shutdown prior to it shutting down

**What is the new behavior (if this is a feature change)?**
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

- The pgBackRest dedicated repo host now waits for all PG instances to shutdown prior to shutting down

**Other Information**:

N/A

### CODE REVIEW

File: internal/controller/postgrescluster/pgbackrest_test.go [hunk 1]
```diff
@@ -2187,6 +2188,34 @@ func TestGenerateRepoHostIntent(t *testing.T) {
...
+				Shutdown: initialize.Bool(true),
+			},
+		}
+		observed := &observedInstances{forCluster: []*Instance{{Pods: []*corev1.Pod{{}}}}}
+		sts, err := r.generateRepoHostIntent(cluster, "", &RepoResources{}, observed)
+		assert.NilError(t, err)
+		assert.Equal(t, *sts.Spec.Replicas, int32(1))
+	})
+
+	t.Run("PG instances observed, shutdown repo host", func(t *testing.T) {
```
Comment (tjmoore4) [inline internal/controller/postgrescluster/pgbackrest_test.go:2208]: :question: Should this be "No PG instances observed...."?

**Review by tjmoore4 — COMMENTED** [review]
:+1: Looks good, one question.

File: internal/controller/postgrescluster/instance.go [hunk 2]
```diff
@@ -1695,12 +1698,23 @@ func (r *Reconciler) prepareForUpgrade(ctx context.Context,
 			ConditionPGUpgradeCompleted)
 	}
 
+	var repoHostReady bool
+	repoHostReadyCondition := meta.FindStatusCondition(cluster.Status.Conditions,
+		ConditionRepoHostReady)
+	if repoHostReadyCondition != nil && repoHostReadyCondition.Status == metav1.ConditionTrue {
+		repoHostReady = true
+	}
```
Comment (cbandy) [inline internal/controller/postgrescluster/instance.go:1706]: ⛏️ There's a function for this, `IsStatusConditionTrue`

```suggestion
	repoHostReady := meta.IsStatusConditionTrue(cluster.Status.Conditions,
		ConditionRepoHostReady)
```

**Review by cbandy — APPROVED** [review]

File: internal/controller/postgrescluster/pgbackrest_test.go (hunk 1 above)
Comment (andrewlecuyer) [reply internal/controller/postgrescluster/pgbackrest_test.go]: Updated to "No PG instances observed".

File: internal/controller/postgrescluster/instance.go (hunk 2 above)
Comment (andrewlecuyer) [reply internal/controller/postgrescluster/instance.go]: :exploding_head: Thanks for the tip!  I know this helper can be leveraged elsewhere as well.

**Review by tjmoore4 — APPROVED** [review]

**Pull request merged**